### DISCUSSION TITLE
Question: feasibility of provisioning Kubernetes cluster on Hetzner with `terranix` and `terranix-hcloud`

### DISCUSSION THREAD

**Discussion by peterbecich** [terranix/terranix#80] [Q&A] [answered]
Disclaimer: I could be misunderstanding these projects.

The [`kubernix` project](https://github.com/saschagrunert/kubernix) exists for setting up a Kubernetes cluster by Nix configuration.

These exist for setting up a Kubernetes cluster on a NixOS system:
- https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/cluster/kubernetes/default.nix
- https://github.com/justinas/nixos-ha-kubernetes

The [`terraform-hcloud-kube-hetzner` project](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner), which does not use Nix at all:
- provisions servers on Hetzner using Terraform
- sets up a Kubernetes (k3s) cluster on those servers
- the OS of these servers is openSUSE

Question:

Is it feasible to combine `terranix` and `terranix-hcloud` with `kubernix`, `nixos-ha-kubernetes` or the official NixOS Kubernetes module to reproduce the features of `terraform-hcloud-kube-hetzner`?

Instead of openSUSE, NixOS would be the OS on each Hetzner machine.

---------------------

Side note:
The [`kubenix` project](https://github.com/hall/kubenix) exists for generating Kubernetes JSON manifests (not YAML manifests) from Nix configuration. I mention this separately because it appears to be outside the scope of `terraform-hcloud-kube-hetzner`. I am primarily interested in the providing the features of `terraform-hcloud-kube-hetzner` in a Nix Flake.

-------------------

Thanks for your time

**Comment by mdarocha** [comment, marked as answer]
You could definitely implement what you propose. To manage NixOS configurations and deployments with Terraform (which terranix is just an interface for), you can check out the approach outlined [here](https://nix.dev/tutorials/nixos/deploying-nixos-using-terraform)
